pkg/skaffold/verify/k8sjob/verify.go

```diff
@@ -0,0 +1,461 @@
+		defaultNamespace, err = util.ExpandEnvTemplate(*v.DefaultNamespace, nil)
+		if err != nil {
+			return nil, err
+		}
+	}
+	kubectl := kubectl.NewCLI(cfg, v.Flags, defaultNamespace)
+	// default namespace must be "default" not "" when used to create and stream logs from Job(s)
+	if defaultNamespace == "" {
+		defaultNamespace = "default"
+	}
+	tracker := tracker.NewContainerTracker()
+	logger, err := k8sjoblogger.NewLogger(ctx, tracker, labeller, kubectl.KubeContext)
+	if err != nil {
+		return nil, err
+	}
+	var origImages []graph.Artifact
+	for _, artifact := range artifacts {
+		origImages = append(origImages, graph.Artifact{
+			ImageName: artifact.ImageName,
+		})
+	}
+	return &Verifier{
+		originalImages:   origImages,
+		cfg:              v.KubernetesClusterVerifyTestCases,
+		defaultNamespace: defaultNamespace,
+		imageLoader:      component.NewImageLoader(cfg, kubectl.CLI),
+		logger:           logger,
+		statusMonitor:    &status.NoopMonitor{},
+		kubectl:          kubectl,
+		tracker:          tracker,
+		labeller:         labeller,
+	}, nil
+}
+
+func (v *Verifier) ConfigName() string {
+	return v.configName
+}
+
+func (v *Verifier) GetLogger() log.Logger {
+	return v.logger
+}
+
+func (v *Verifier) GetStatusMonitor() status.Monitor {
+	return v.statusMonitor
+}
+
+func (v *Verifier) RegisterLocalImages(images []graph.Artifact) {
+	v.localImages = images
+}
+
+func (v *Verifier) TrackBuildArtifacts(artifacts []graph.Artifact) {
+	v.logger.RegisterArtifacts(artifacts)
+}
+
+// Verify executes specified artifacts by creating containers in the local docker daemon
+// from each specified image, executing them, and waiting for execution to complete.
+func (v *Verifier) Verify(ctx context.Context, out io.Writer, allbuilds []graph.Artifact) error {
+	var (
+		childCtx context.Context
+		endTrace func(...trace.SpanEndOption)
+		wg       sync.WaitGroup
+	)
+
+	// TODO(aaron-prindle) add trace info
+	if err := kubernetes.FailIfClusterIsNotReachable(v.kubectl.KubeContext); err != nil {
+		return fmt.Errorf("unable to connect to Kubernetes: %w", err)
+	}
+
+	childCtx, endTrace = instrumentation.StartTrace(ctx, "Verify_LoadImages")
+	if err := v.imageLoader.LoadImages(childCtx, out, v.localImages, v.originalImages, allbuilds); err != nil {
+		endTrace(instrumentation.TraceEndError(err))
+		return err
+	}
+	endTrace()
+
+	builds := []graph.Artifact{}
+	const maxWorkers = math.MaxInt64
+	s := semgroup.NewGroup(context.Background(), maxWorkers)
+
+	for _, tc := range v.cfg {
+		var na graph.Artifact
+		foundArtifact := false
+		nTC := *tc
+
+		for _, b := range allbuilds {
+			if tc.Container.Image == b.ImageName {
+				foundArtifact = true
+				na = graph.Artifact{
+					ImageName: tc.Container.Image,
+					Tag:       b.Tag,
+				}
+				builds = append(builds, graph.Artifact{
+					ImageName: tc.Container.Image,
+					Tag:       b.Tag,
+				})
+				nTC.Container.Image = na.Tag
+				break
+			}
+		}
+		if !foundArtifact {
+			na = graph.Artifact{
```

Comment:
Not sure if I understood the `na` variable correctly, but does it looks like we are not using it in here?

Reply:
done, removed na entirely